File: examples/java/Shapes.java
package com.example.demo;

/**
 * Shape is something with a measurable area.
 */
public interface Shape {

    /** area returns the shape's area in square units. */
    double area();

    String name();
}

/**
 * Point is an immutable 2D coordinate.
 */
record Point(int x, int y) {
}

/**
 * Circle is a Shape defined by its radius.
 */
class Circle implements Shape {

    private final double radius;

    Circle(double radius) {
        this.radius = radius;
    }

    @Override
    public double area() {
        return Math.PI * radius * radius;
    }

    @Override
    public String name() {
        return "circle";
    }
}

/**
 * Town greets visitors. Its greet method is a deliberate homonym of
 * Greeter.greet, with a different container and a different signature, to
 * exercise SymbolID disambiguation.
 */
class Town {

    public String greet() {
        return "welcome to town";
    }
}

File: examples/typescript/shapes.ts
/**
 * Shape is anything with a measurable area.
 * @returns the area when area() is called
 */
interface Shape {
  /** area returns the shape's area. */
  area(): number;
  name: string;
}

/** Circle is a round shape. */
class Circle implements Shape {
  name: string;

  constructor(private r: number) {
    this.name = "circle";
  }

  /** area returns the circle's area. */
  area(): number {
    return 3.14159 * this.r * this.r;
  }

  /** unit builds the unit circle. */
  static unit(): Circle {
    return new Circle(1);
  }
}

/** Square is a four-sided shape. */
class Square implements Shape {
  name: string;

  constructor(private side: number) {
    this.name = "square";
  }

  area(): number {
    return this.side * this.side;
  }
}

// area is a free function sharing its name with the methods above.
function area(w: number, h: number): number {
  return w * h;
}
